exit_gtk.py

```diff
@@ -0,0 +1,222 @@
+import pygtk
+pygtk.require('2.0')
+import gtk
+import gobject
+import logging
+import collections
+from colored_image_button import ColoredImageButton
+import struct
+from time import sleep
+import dbus_interface
+exit_log = logging.getLogger('Bunsen-Exit-Log')
+
+
+class ExitGtk:
+	color_button = None
+	window = None
+	#get a DBusInterface instance, so we can send message out.
+	exit_bus = dbus_interface.DbusInterface()
+
+	def _key_press_event(self, widget, event):
+		keyval = event.keyval
+		keyval_name = gtk.gdk.keyval_name(keyval)
+		state = event.state
+		alt = (state & gtk.gdk.MOD1_MASK)
+		# Cancel shortcut
+		if alt and keyval_name == 'l':
+			self.send_to_dbus('Logout')
+		# Suspend shortcut
+		elif alt and keyval_name == 's':
+			self.send_to_dbus('Suspend')
+		# Reboot Shortcut
+		elif alt and keyval_name == 'b':
+			self.send_to_dbus('Reboot')
+		# Poweroff shortcut
+		elif alt and keyval_name == 'p':
+			self.send_to_dbus('PowerOff')
+		# Hibernate Shortcut
+		elif alt and keyval_name == 'i':
+			self.send_to_dbus('Hibernate')
+		# Hybrid Sleep Shortcut
+		elif alt and keyval_name == 'y':
+			self.send_to_dbus('HybridSleep')
+		# Cancel Shortcut
+		elif keyval_name == 'Escape':
+			self.destroy()
+		elif keyval_name == 'Alt':
+			self.show_button_labels()
+		else:
+			return False
+		return True
+		
+	def destroy(self, widget=None, event=None, data=None):
+		self.window.hide_all()
+		gtk.main_quit()
+
+	def __init__(self, button_values, exit_bus, theme, theme_entries, style_path):
+		# Attributes
+		self.theme_entries = theme_entries
+		self.window = gtk.Window(gtk.WINDOW_TOPLEVEL)
+		try:
+			self.dialog_height = int(self.theme_entries['dialog_height'])
+		except:
+			exit_log.warn("dialog_height not set. Settins a default of 64 pixels.")
+			self.dialog_height = 64
+		# Get the screen width under the cursor
+		screen_width = 800 # fallback width
+		try:
+			display=gtk.gdk.Display(gtk.gdk.get_display())
+			screen, x, y, flags=display.get_pointer()
+			curmon = screen.get_monitor_at_point(x, y)
+			_, _, screen_width, _ = screen.get_monitor_geometry(curmon)
+		except:
+			exit_log.info('Error constructing UI. Not running under X')
+		finally:
+			del x, y, display, screen, curmon
+		exit_log.debug("Detected screen_width is " + str(screen_width))
+		try:
+			self.width_adjustment = float(self.theme_entries['window_width_adjustment'])
+		except:
+			exit_log.warn("window_width_adjustment not a float. Please check config. Defaulting to 0.5")
```

Comment:
Missing `self.width_adjustment = 0.5`, leading to segfault for me later on.

Reply:
If no configuration can be found for whatever reason, you should now get a dialog built using the default gtk theme for the system. If one can be found, then testing that each value is valid and if not, set to some sane default is still a work-in-progress. Also, the configuration files are still in a state of flux due to the fact that I have been mapping in functionality as I go along. However, at the moment, the program is looking for bl-exitrc in ~/.config/bunsen-exit. 

I noticed on the custom buttons, that the enter key was not being handled such that tab...tab..enter was activating the button as one would expect. The lastest commit address that as well. I will be working on this pretty heavily for the next couple of days so expect more incoming commits.

Reply:
Yup, now it works. Thanks.

Reply:
This looks excellent - thank you!

> the configuration files are still in a state of flux due to the fact that I have been mapping in functionality as I go along. However, at the moment, the program is looking for bl-exitrc in ~/.config/bunsen-exit.

I'm wondering about backwards compatibility here. Is there any possibility of having old-style bl-exitrc files interpreted to some extent? (The same philosophy you applied with falling back on a default style when there's no config file at all.)

Reply:
I have been planning to do some refactoring on all of the code that tests for valid configuration entries and sets them to some sane default, quite possibly the stock BL look for Hydrogen, since that is where we started and the old #! look would be better covered by using a #! gtk theme and no config file at all.  At the moment, those tests are scattered all throughout the program. I want to group them all up and bring them back under the exit_config module.py. With these tests, if an option cannot be parsed, a sensible default will be set and if logging is enabled, an entry will be written with a [WARN] tag.

So my plan is to group the logging levels as such:
[INFO] General and fairly brief info, such as logging started, which configuration is being loaded, etc.
[WARN] This logging level would be for modders. A warning entry means that a configuration item could not be parsed and so the program is setting some default as I mentioned above. This would be our way of alerting those who wish to mod the program, "Hey I tried to parse this configuration item but could not so I set it to something meaningful."
[DEBUG] This level is basically full introspection of all the inputs and outputs of the program. It is very verbose but gives a good view of what is going on in the program.

In the case that the program attempts to parse an old bl-exitrc, the button entries have changed. The old style looks like this:
```
[cancel]
show = always
onerror = novisual

[logout]
show = always
onerror = novisual

[suspend]
show = always
onerror = novisual

[sleep]
show = always
onerror = novisual

[hibernate]
show = never
onerror = novisual

[hybridsleep]
show = never
onerror = novisual

[reboot]
show = always
onerror = novisual

[poweroff]
show = always
onerror = novisual
```
While the new one looks like this:
```
[button_values]
Cancel = show
Logout = show
Suspend = show
Reboot = show
PowerOff = show
Hibernate = show
HybridSleep = show
`````
The onerror stuff has been replaced by a different logging mechanism. Additionally, I don't pop up error dialogs anywhere. You remember the Windows donk with the blocking dialog ala Windows 95 - 98? Yeah, I hate those things.

What should happen is that the program will decide it can't use the button array as presented in the old bl-exitrc and build a default button list which shows every button except Hibernate and HybridSleep by default. I can place a log message that tells the user is cannot parse the button array, most likely because it is using an old configuration format and set the default array, while encouraging them to update their configuration file.

Similarly, if the theme specified cannot be found, it will build an exit dialog based on some set of defaults (I mentioned Hydrogen's look earlier.) This behavior will continue for each theme_entry in the specified theme, each time issuing a [WARN] in the logs if it overrides some theme_entry.

To summarize, I can ensure that the program will build some sort of default themed exit dialog if it finds an old style bl-exitrc. I can also issue [WARN] logs that encourage folks to use the new configuration setup. The new configuration setup, I hope, can become a tool that our modding community can use in their desktop builds. There is additional functionality that I have not mapped in yet but is possible, most notably, the ability to build button borders and the ability to set fonts and their associated sizes for the button labels. 

Finally, I will ensure that the new configuration files end up heavily commented so that it will hopefully be apparent to folks how to use the config files to build their own exit dialog themes.

I hope this will serve to alleviate your concerns somewhat.

Reply:
> I hope this will serve to alleviate your concerns somewhat.

Yes, indeed. Thank you!